### PR TITLE
Implemented css animations

### PR DESCRIPTION
**Pull request by tzraikov**
#1608

### CODE REVIEW

**Comment by dtopuzov** [comment]
run ci

**Comment by dtopuzov** [comment]
run ci

**Comment by hdeshev** [comment]
We need tests here. CSS parsing, and transform/keyframe extraction can probably be tested without any UI whatsoever.

Also, please rebase on top of master to get rid of the merge commits and squash unnecessary commits.

**Comment by PanayotCankov** [comment]
Squash and rebase onto master, please.

File: tsconfig.json [hunk 1]
```diff
@@ -90,10 +90,16 @@
         "apps/perf-tests/ApplicationLoadTimeAndFPS/mainPage.ts",
         "apps/perf-tests/ApplicationSize/app.ts",
         "apps/perf-tests/ApplicationSize/mainPage.ts",
+        "apps/perf-tests/common.d.ts",
+        "apps/perf-tests/common.ts",
         "apps/perf-tests/ComplexObjectGraphMemoryTest/app.ts",
         "apps/perf-tests/ComplexObjectGraphMemoryTest/mainPage.ts",
         "apps/perf-tests/ControlCreationSpeedTest/app.ts",
         "apps/perf-tests/ControlCreationSpeedTest/mainPage.ts",
+        "apps/perf-tests/controls-page.d.ts",
```
Comment (PanayotCankov) [inline tsconfig.json]: Some of the files have been just shuffles around and are hard to track. Can you restore the original order so we can see here easily insertions and deletions?

File: ui/animation/animation.android.ts [hunk 2]
```diff
@@ -165,7 +171,15 @@ export class Animation extends common.Animation implements definition.Animation
                 originalValue1 = nativeView.getAlpha();
                 nativeArray = (<any>Array).create("float", 1);
                 nativeArray[0] = propertyAnimation.value;
-                propertyUpdateCallbacks.push(checkAnimation(() => { propertyAnimation.target.opacity = propertyAnimation.value }));
+                if (this._valueSource !== undefined) {
+                    propertyUpdateCallbacks.push(checkAnimation(() => { 
+                        propertyAnimation.target.style._resetValue(styleModule.opacityProperty);
```
Comment (PanayotCankov) [inline ui/animation/animation.android.ts]: I guess _setValue wont work without _resetValue? Why?

File: ui/animation/animation.android.ts [hunk 3]
```diff
@@ -237,10 +269,20 @@ export class Animation extends common.Animation implements definition.Animation
                 originalValue1 = nativeView.getScaleX();
                 originalValue2 = nativeView.getScaleY();
 
-                propertyUpdateCallbacks.push(checkAnimation(() => {
-                    propertyAnimation.target.scaleX = propertyAnimation.value.x;
-                    propertyAnimation.target.scaleY = propertyAnimation.value.y;
-                }));
+                if (this._valueSource !== undefined) {
```
Comment (PanayotCankov) [inline ui/animation/animation.android.ts]: These if-s check what the ValueSource is and set either the property on the style, or as local value on the object.
Can't we use the: https://github.com/NativeScript/NativeScript/blob/master/ui/core/dependency-observable.d.ts#L222
overload with the source arg, and have [CSS](https://github.com/NativeScript/NativeScript/blob/master/ui/core/dependency-observable.d.ts#L283) or [Local](https://github.com/NativeScript/NativeScript/blob/master/ui/core/dependency-observable.d.ts#L287) by default?
The if-s just look redundant.

File: ui/animation/animation.android.ts [hunk 4]
```diff
@@ -302,11 +352,11 @@ export class Animation extends common.Animation implements definition.Animation
     }
 }
 
-var easeIn = new android.view.animation.AccelerateInterpolator(1);
-var easeOut = new android.view.animation.DecelerateInterpolator(1);
-var easeInOut = new android.view.animation.AccelerateDecelerateInterpolator();
-var linear = new android.view.animation.LinearInterpolator();
-var bounce = new android.view.animation.BounceInterpolator();
+let easeIn = new android.view.animation.AccelerateInterpolator(1);
```
Comment (PanayotCankov) [inline ui/animation/animation.android.ts]: Probably the migration from `var`s to `let`s had to happen in a separate PR.

File: ui/animation/animation.ios.ts [hunk 5]
```diff
@@ -29,42 +29,70 @@ class AnimationDelegateImpl extends NSObject {
...
-               }
-               if (value[common.Properties.scale] !== undefined) {
-                  this._propertyAnimation.target.scaleX = value[common.Properties.scale].x;
-                  this._propertyAnimation.target.scaleY = value[common.Properties.scale].y;
-               }
-               break;
-        }
-
-        (<any>this._propertyAnimation.target)._resumePresentationLayerUpdates();
+       if (this._valueSource !== undefined) {
```
Comment (PanayotCankov) [inline ui/animation/animation.ios.ts]: Same redundant if here. Cant we set the `this._valueSource = valueSource || ValueSource.Local` in the constructor and avoid the branch here.

File: ui/animation/animationgroup.d.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        keyframes: Array<Keyframe>;
+
+        /**
+         * Plays the animation.
+         */
+        public play: (view: view.View) => Promise<void>;
+
+        public static animationGroupFromSelectorDeclarations(declarations: cssParser.Declaration[]): AnimationGroup;
+
+        public static keyframesFromCSS(cssKeyframes: Object): Array<Keyframe>;
```
Comment (PanayotCankov) [inline ui/animation/animationgroup.d.ts]: Can we move this factory method in the module in which it is used, or in a separate module. The object model for key-frame animations doesn't necessary require the css parser.

File: ui/animation/animationgroup.ts [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+
+    public static keyframesFromCSS(cssKeyframes: Object): Array<Keyframe> {
+        let parsedKeyframes = new Array<Keyframe>();
+        for (let keyframe of (<any>cssKeyframes).keyframes) {
+            let declarations = AnimationGroup.parseKeyframeDeclarations(keyframe);
+            for (let time of keyframe.values) {
+                if (time === "from") {
+                    time = 0;
+                }
+                if (time === "to") {
```
Comment (PanayotCankov) [inline ui/animation/animationgroup.ts]: `} else if {`?

File: ui/animation/animationgroup.ts [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+    }
+
+    public static animationGroupFromSelectorDeclarations(declarations: cssParser.Declaration[]): AnimationGroup {
+        let animationInfo = undefined;
+        for (let declaration of declarations) {
+            if (declaration.property.indexOf("animation") === 0) {
+                if (animationInfo === undefined) {
+                    animationInfo = new AnimationGroup();
+                }
+                switch (declaration.property) {
```
Comment (PanayotCankov) [inline ui/animation/animationgroup.ts]: Can we make a jump table? Like expose a dictionary in the module:
``` TypeScript
animationProperties = {
  "animation-name": (i, d) => i.name = d.value;
  "animation-duration": (i, d) => i.duration = converters.timeConverter(declaration.value);
  // ...
}
```
And use it here like:
``` TypeScript
var propertyHandler = animationProperties[declaration.property];
if (propertyHandler) {
	propertyHandler(animationInfo, declaration);
} else {
    // unknown property
}
```

File: ui/styling/css-selector.ts [hunk 9]
```diff
@@ -51,19 +54,49 @@ export class CssSelector {
...
             try {
-                view.style._setValue(property, value, observable.ValueSource.Css);
+                view.style._setValue(property, value, modifier);
             }
             catch (ex) {
                 trace.write("Error setting property: " + property.name + " view: " + view + " value: " + value + " " + ex, trace.categories.Style, trace.messageType.error);
             }
         });
+        if (this.isAnimated) {
+            if (this._animationGroup.isPlaying) {
```
Comment (PanayotCankov) [inline ui/styling/css-selector.ts]: The selector should be able to play with multiple view instances. For example if a list view get 3 items selected out of 10 and has keyframe animation tinting selected items it should animate all of the 3. Perhaps the AnimationGroup should not have `isPlaying` but rather the "play" method of the AnimationGroup should return a somewhat token used to play/pause/cancel, check for progress/isPlaying etc.

**Comment by PanayotCankov** [comment]
:+1: I think we should mark this with API stability index, merge it and start the dogfooding.

**Comment by tzraikov** [comment]
run ci

**Pull request closed without merging**